compiler/noirc_evaluator/src/ssa/opt/inline_functions_with_at_most_one_instruction.rs

```diff
@@ -104,6 +111,69 @@ mod test {
         ");
     }
 
+    #[test]
+    fn does_not_inline_functions_that_require_multiple_passes() {
+        let src = "
+        acir(inline) fn main f0 {
+          b0(v0: Field):
+            v1 = call f2(v0) -> Field
+            return v1
+        }
+
+        acir(inline) fn foo f1 {
+          b0(v0: Field):
+            return v0
+        }
+
+        acir(inline) fn bar f2 {
+          b0(v0: Field):
+            v1 = call f1(v0) -> Field
+            v2 = call f1(v0) -> Field
+            v3 = add v1, v2
+            return v3
+        }
+        ";
+        let ssa = Ssa::from_str(src).unwrap();
+
+        // In the first pass it won't recognize that `main` could be simplified.
+        let ssa = ssa.inline_functions_with_at_most_one_instruction();
+        assert_ssa_snapshot!(ssa.clone(), @r"
+        acir(inline) fn main f0 {
+          b0(v0: Field):
+            v2 = call f2(v0) -> Field
+            return v2
+        }
+        acir(inline) fn foo f1 {
+          b0(v0: Field):
+            return v0
+        }
+        acir(inline) fn bar f2 {
+          b0(v0: Field):
+            v1 = add v0, v0
+            return v1
+        }
+        ");
+
+        // After `bar` has been simplified, it does `main` as well.
+        let ssa = ssa.inline_functions_with_at_most_one_instruction();
```

Comment:
```suggestion
        let ssa = Ssa::from_str(src).unwrap();
        // After `bar` has been simplified, it does `main` as well.
        let ssa = ssa
            .inline_functions_with_at_most_one_instruction()
            .inline_functions_with_at_most_one_instruction();
```
let's just regenerate the SSA rather than cloning it

Reply:
Okay. Curious though, what problem do you see in having a `Ssa::clone`? After all the SSA can be serialised and deserialised, it can be printed and parsed, all of which provide roundabout ways to clone it. Originally I wanted to capture the subsequent states of it in https://github.com/noir-lang/noir/pull/8393, but ended up just using the string representation. I'll remove this commit from that PR as well if it's considered harmful.

Reply:
I'm not very against it being used in tests like how you did, but if we do not need to clone I think it is good to discourage cloning the entire Ssa structure. Thus, not deriving clone on `Ssa`. 
- Cloning a large Ssa can potentially be very expensive if it works its way into our normal compilation pipeline
- For our central Ssa data structure it is good to enforce ownership tracking. Cloning could let us avoid that friction more easily, but we don't really want that

> SSA can be serialised and deserialised, it can be printed and parsed, all of which provide roundabout ways to (partially) clone it

Yes, but as you said they are roundabout ways to perform a clone as it is not the actual desired functionality in each of those cases.

Reply:
Okay, thanks 👍 
Found a way to avoid the duplication using `&mut` with the macro, hopefully best of both worlds.